File: server/keycloak.ts
import jwt from 'jsonwebtoken';
import jwksClient from 'jwks-rsa';
import { logger } from './logger.js';
import { CircuitBreaker } from './services/circuit-breaker.js';

const KEYCLOAK_URL = process.env.KEYCLOAK_URL || 'http://localhost:8080';
const KEYCLOAK_REALM = process.env.KEYCLOAK_REALM || 'farmer-realm';
const KEYCLOAK_CLIENT_ID = process.env.KEYCLOAK_CLIENT_ID || 'farmer-api';
const KEYCLOAK_CLIENT_SECRET = process.env.KEYCLOAK_CLIENT_SECRET;

const keycloakBreaker = new CircuitBreaker({ name: 'keycloak', failureThreshold: 5, resetTimeoutMs: 30_000, timeoutMs: 8_000 });

// Cached service account token
let _cachedServiceToken: string | null = null;
let _tokenExpiresAt = 0;

// JWKS client for token verification
const jwksClientInstance = jwksClient({
  jwksUri: `${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}/protocol/openid-connect/certs`,
  cache: true,
  cacheMaxAge: 86400000, // 24 hours
  rateLimit: true,
  jwksRequestsPerMinute: 10,
});

/**
 * Get signing key from Keycloak JWKS
 */
function getKey(header: jwt.JwtHeader, callback: jwt.SigningKeyCallback) {
  jwksClientInstance.getSigningKey(header.kid, (err, key) => {
    if (err) {
      callback(err);
      return;
    }
    const signingKey = key?.getPublicKey();
    callback(null, signingKey);
  });
}

/**
 * Keycloak user interface
 */
export interface KeycloakUser {
  id: string;
  email: string;
  firstName?: string;
  lastName?: string;
  roles: string[];
  username: string;
}

/**
 * Verify Keycloak JWT token
 */
export async function verifyKeycloakToken(token: string): Promise<KeycloakUser | null> {
  return new Promise((resolve) => {
    jwt.verify(
      token,
      getKey,
      {
        algorithms: ['RS256'],
        issuer: `${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}`,
        audience: KEYCLOAK_CLIENT_ID,
      },
      (err, decoded) => {
        if (err) {
          logger.warn('[Keycloak] Token verification failed', { error: err.message });
          resolve(null);
          return;
        }

        if (!decoded) {
          resolve(null);
          return;
        }

        const payload = decoded as Record<string, unknown>;
        const realmAccess = payload.realm_access as { roles?: string[] } | undefined;
        const user: KeycloakUser = {
          id: String(payload.sub || ''),
          email: String(payload.email || payload.preferred_username || ''),
          firstName: String(payload.given_name || ''),
          lastName: String(payload.family_name || ''),
          username: String(payload.preferred_username || payload.email || ''),
          roles: realmAccess?.roles || [],
        };

        resolve(user);
      }
    );
  });
}

/**
 * Get service account token for backend-to-backend calls
 */
export async function getServiceAccountToken(): Promise<string | null> {
  if (!KEYCLOAK_CLIENT_SECRET) {
    logger.warn('[Keycloak] Client secret not configured');
    return null;
  }

  // Return cached token if still valid (with 30s buffer)
  if (_cachedServiceToken && Date.now() < _tokenExpiresAt - 30_000) {
    return _cachedServiceToken;
  }

  try {
    const response = await keycloakBreaker.execute(() =>
      fetch(
        `${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}/protocol/openid-connect/token`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          body: new URLSearchParams({
            grant_type: 'client_credentials',
            client_id: KEYCLOAK_CLIENT_ID,
            client_secret: KEYCLOAK_CLIENT_SECRET,
          }),
          signal: AbortSignal.timeout(8000),
        }
      )
    );

    if (!response.ok) {
      logger.warn('[Keycloak] Failed to get service account token', { status: response.status });
      return null;
    }

    const data = (await response.json()) as { access_token: string; expires_in: number };
    _cachedServiceToken = data.access_token;
    _tokenExpiresAt = Date.now() + data.expires_in * 1000;
    return data.access_token;
  } catch (error) {
    logger.error('[Keycloak] Error getting service account token', { error: (error as Error).message });
    return null;
  }
}

/**
 * Introspect token (validate and get user info)
 */
export async function introspectToken(token: string): Promise<Record<string, unknown> | null> {
  if (!KEYCLOAK_CLIENT_SECRET) {
    logger.warn('[Keycloak] Client secret not configured');
    return null;
  }

  try {
    const response = await keycloakBreaker.execute(() =>
      fetch(
        `${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}/protocol/openid-connect/token/introspect`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          body: new URLSearchParams({
            token,
            client_id: KEYCLOAK_CLIENT_ID,
            client_secret: KEYCLOAK_CLIENT_SECRET,
          }),
          signal: AbortSignal.timeout(5000),
        }
      )
    );

    if (!response.ok) {
      logger.warn('[Keycloak] Token introspection failed', { status: response.status });
      return null;
    }

    const data = (await response.json()) as { active: boolean; [key: string]: unknown };
    return data.active ? data : null;
  } catch (error) {
    logger.error('[Keycloak] Error introspecting token', { error: (error as Error).message });
    return null;
  }
}

/**
 * Get user info from Keycloak
 */
export async function getUserInfo(token: string): Promise<Record<string, unknown> | null> {
  try {
    const response = await keycloakBreaker.execute(() =>
      fetch(
        `${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}/protocol/openid-connect/userinfo`,
        {
          headers: { Authorization: `Bearer ${token}` },
          signal: AbortSignal.timeout(5000),
        }
      )
    );

    if (!response.ok) {
      logger.warn('[Keycloak] Failed to get user info', { status: response.status });
      return null;
    }

    return (await response.json()) as Record<string, unknown>;
  } catch (error) {
    logger.error('[Keycloak] Error getting user info', { error: (error as Error).message });
    return null;
  }
}

export function isKeycloakHealthy(): boolean {
  return keycloakBreaker.getState().state !== 'OPEN';
}

/**
 * Check if user has required role
 */
export function hasRole(user: KeycloakUser | null, role: string): boolean {
  if (!user) return false;
  return user.roles.includes(role);
}

/**
 * Check if user has any of the required roles
 */
export function hasAnyRole(user: KeycloakUser | null, roles: string[]): boolean {
  if (!user) return false;
  return roles.some(role => user.roles.includes(role));
}

/**
 * Check if user has all required roles
 */
export function hasAllRoles(user: KeycloakUser | null, roles: string[]): boolean {
  if (!user) return false;
  return roles.every(role => user.roles.includes(role));
}

/**
 * Keycloak configuration info
 */
export const keycloakConfig = {
  url: KEYCLOAK_URL,
  realm: KEYCLOAK_REALM,
  clientId: KEYCLOAK_CLIENT_ID,
  hasClientSecret: !!KEYCLOAK_CLIENT_SECRET,
};

logger.info('[Keycloak] Configuration loaded', { url: KEYCLOAK_URL, realm: KEYCLOAK_REALM, clientId: KEYCLOAK_CLIENT_ID, hasSecret: !!KEYCLOAK_CLIENT_SECRET });

File: services/go/shared/middleware/keycloak.go
package middleware

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// KeycloakUser represents a user from Keycloak
type KeycloakUser struct {
	ID        string   `json:"id"`
	Email     string   `json:"email"`
	FirstName string   `json:"first_name,omitempty"`
	LastName  string   `json:"last_name,omitempty"`
	Username  string   `json:"username"`
	Roles     []string `json:"roles"`
}

// KeycloakConfig holds Keycloak configuration
type KeycloakConfig struct {
	URL          string
	Realm        string
	ClientID     string
	ClientSecret string
}

// KeycloakClient provides Keycloak authentication operations
type KeycloakClient struct {
	config     KeycloakConfig
	httpClient *http.Client
	jwksCache  map[string]interface{}
	jwksMu     sync.RWMutex
	jwksExpiry time.Time
}

// NewKeycloakClient creates a new Keycloak client
func NewKeycloakClient(config KeycloakConfig) *KeycloakClient {
	if config.URL == "" {
		config.URL = "http://localhost:8080"
	}
	if config.Realm == "" {
		config.Realm = "farmer-realm"
	}
	if config.ClientID == "" {
		config.ClientID = "farmer-api"
	}

	return &KeycloakClient{
		config: config,
		httpClient: &http.Client{
			Timeout: 10 * time.Second,
		},
		jwksCache: make(map[string]interface{}),
	}
}

// VerifyToken verifies a JWT token and returns the user
func (c *KeycloakClient) VerifyToken(ctx context.Context, tokenString string) (*KeycloakUser, error) {
	// Parse the token
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		// Verify signing method
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}

		// Get the key ID
		kid, ok := token.Header["kid"].(string)
		if !ok {
			return nil, fmt.Errorf("missing key ID in token header")
		}

		// Get the public key from JWKS
		return c.getPublicKey(ctx, kid)
	})

	if err != nil {
		log.Printf("[Keycloak] Token verification failed: %v", err)
		return nil, err
	}

	if !token.Valid {
		return nil, fmt.Errorf("invalid token")
	}

	// Extract claims
	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return nil, fmt.Errorf("invalid token claims")
	}

	// Verify issuer
	expectedIssuer := fmt.Sprintf("%s/realms/%s", c.config.URL, c.config.Realm)
	if iss, _ := claims["iss"].(string); iss != expectedIssuer {
		return nil, fmt.Errorf("invalid issuer: expected %s, got %s", expectedIssuer, iss)
	}

	// Extract user information
	user := &KeycloakUser{
		ID:       claims["sub"].(string),
		Username: getStringClaim(claims, "preferred_username"),
		Email:    getStringClaim(claims, "email"),
	}

	if user.Email == "" {
		user.Email = user.Username
	}

	user.FirstName = getStringClaim(claims, "given_name")
	user.LastName = getStringClaim(claims, "family_name")

	// Extract roles from realm_access
	if realmAccess, ok := claims["realm_access"].(map[string]interface{}); ok {
		if roles, ok := realmAccess["roles"].([]interface{}); ok {
			for _, role := range roles {
				if roleStr, ok := role.(string); ok {
					user.Roles = append(user.Roles, roleStr)
				}
			}
		}
	}

	return user, nil
}

// getPublicKey retrieves the public key from JWKS
func (c *KeycloakClient) getPublicKey(ctx context.Context, kid string) (interface{}, error) {
	c.jwksMu.RLock()
	if key, ok := c.jwksCache[kid]; ok && time.Now().Before(c.jwksExpiry) {
		c.jwksMu.RUnlock()
		return key, nil
	}
	c.jwksMu.RUnlock()

	// Fetch JWKS
	jwksURL := fmt.Sprintf("%s/realms/%s/protocol/openid-connect/certs", c.config.URL, c.config.Realm)
	resp, err := c.httpClient.Get(jwksURL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch JWKS: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("JWKS request failed with status: %d", resp.StatusCode)
	}

	var jwks struct {
		Keys []map[string]interface{} `json:"keys"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&jwks); err != nil {
		return nil, fmt.Errorf("failed to decode JWKS: %w", err)
	}

	// Cache the keys
	c.jwksMu.Lock()
	c.jwksExpiry = time.Now().Add(24 * time.Hour)
	for _, key := range jwks.Keys {
		if keyID, ok := key["kid"].(string); ok {
			c.jwksCache[keyID] = key
		}
	}
	c.jwksMu.Unlock()

	// Return the requested key
	c.jwksMu.RLock()
	defer c.jwksMu.RUnlock()
	if key, ok := c.jwksCache[kid]; ok {
		return key, nil
	}

	return nil, fmt.Errorf("key not found: %s", kid)
}

// GetServiceAccountToken gets a service account token for backend-to-backend calls
func (c *KeycloakClient) GetServiceAccountToken(ctx context.Context) (string, error) {
	if c.config.ClientSecret == "" {
		return "", fmt.Errorf("client secret not configured")
	}

	tokenURL := fmt.Sprintf("%s/realms/%s/protocol/openid-connect/token", c.config.URL, c.config.Realm)

	data := url.Values{}
	data.Set("grant_type", "client_credentials")
	data.Set("client_id", c.config.ClientID)
	data.Set("client_secret", c.config.ClientSecret)

	req, err := http.NewRequestWithContext(ctx, "POST", tokenURL, strings.NewReader(data.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to get service account token: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("token request failed with status: %d", resp.StatusCode)
	}

	var tokenResp struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&tokenResp); err != nil {
		return "", fmt.Errorf("failed to decode token response: %w", err)
	}

	return tokenResp.AccessToken, nil
}

// IntrospectToken introspects a token to validate and get user info
func (c *KeycloakClient) IntrospectToken(ctx context.Context, token string) (map[string]interface{}, error) {
	if c.config.ClientSecret == "" {
		return nil, fmt.Errorf("client secret not configured")
	}

	introspectURL := fmt.Sprintf("%s/realms/%s/protocol/openid-connect/token/introspect", c.config.URL, c.config.Realm)

	data := url.Values{}
	data.Set("token", token)
	data.Set("client_id", c.config.ClientID)
	data.Set("client_secret", c.config.ClientSecret)

	req, err := http.NewRequestWithContext(ctx, "POST", introspectURL, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to introspect token: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("introspection failed with status: %d", resp.StatusCode)
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode introspection response: %w", err)
	}

	// Check if token is active
	if active, ok := result["active"].(bool); !ok || !active {
		return nil, fmt.Errorf("token is not active")
	}

	return result, nil
}

// GetUserInfo gets user info from the userinfo endpoint
func (c *KeycloakClient) GetUserInfo(ctx context.Context, token string) (map[string]interface{}, error) {
	userInfoURL := fmt.Sprintf("%s/realms/%s/protocol/openid-connect/userinfo", c.config.URL, c.config.Realm)

	req, err := http.NewRequestWithContext(ctx, "GET", userInfoURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to get user info: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("userinfo request failed with status: %d", resp.StatusCode)
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode userinfo response: %w", err)
	}

	return result, nil
}

// HasRole checks if a user has a specific role
func HasRole(user *KeycloakUser, role string) bool {
	if user == nil {
		return false
	}
	for _, r := range user.Roles {
		if r == role {
			return true
		}
	}
	return false
}

// HasAnyRole checks if a user has any of the specified roles
func HasAnyRole(user *KeycloakUser, roles []string) bool {
	if user == nil {
		return false
	}
	for _, role := range roles {
		if HasRole(user, role) {
			return true
		}
	}
	return false
}

// HasAllRoles checks if a user has all of the specified roles
func HasAllRoles(user *KeycloakUser, roles []string) bool {
	if user == nil {
		return false
	}
	for _, role := range roles {
		if !HasRole(user, role) {
			return false
		}
	}
	return true
}

// Helper function to safely get string claims
func getStringClaim(claims jwt.MapClaims, key string) string {
	if val, ok := claims[key].(string); ok {
		return val
	}
	return ""
}
